### DISCUSSION TITLE
Getting an undersampled spiral to reconstruct decently

### DISCUSSION THREAD

**Discussion by nathanjbeaumont** [MRsources/MRzero-Core#79] [Q&A]
Hello,
I have an under sampled spiral signal that I would like to reconstruct. The default settings for reco_adjoint are not outputting the image quality I am trying to accomplish.
This is the k-space:
![image](https://github.com/user-attachments/assets/a76b8434-7e7b-43fe-83e3-298ff3cbfee3)
and here is the reconstruction of the subject05brain phantom:
![image](https://github.com/user-attachments/assets/7baa300a-b9fc-406e-ab71-1afab49cc9b8)

A similar spiral used in a Siemens scanner was able to get this image of a phantom reconstructed with size 256x256 with a sliding window method so I am aiming for something of this quality:
![image](https://github.com/user-attachments/assets/1d360cfd-c27e-4876-90f7-2ce853c9e6b0)

Are there any parameters I could add to the reconstruction method call to accomplish this?

[spiralUndersampled.zip](https://github.com/user-attachments/files/17611404/spiralUndersampled.zip)

**Comment by mzaiss** [comment]
you can try torchkbnufft
https://github.com/mzaiss/MRTwin_pulseq/blob/mr0-core/ex/solF01_bSSFP_2D_radial_torchkbnufft.py

with density correction

you can also try compressed sensing with MIRT or BART, but you have to use multiple reveive coils in the simulation to get similar performance.

**Reply by nathanjbeaumont** [reply]
Thanks, the paper I'm trying to reproduce used a NUFFT so I'll try the torchkbnufft.

Looking into the code, the operation being done in `reco_adjoint` is an FT where the sample points are non-uniform but the "frequency" changes uniformly. (Looking at a single voxel in the `phase` matrix, the phase is changing non uniformly for a spiral, but between voxels the phase changes uniformly.) According to Wikipedia this would be an "adjoint NUFT" is this correct?
https://en.wikipedia.org/wiki/Non-uniform_discrete_Fourier_transform#Definition

**Reply by nathanjbeaumont** [reply]
Unfortunately the recon quality is still unusable with torchkbnufft.
Here it is with the density compensation function being the distance to the origin in k-space:
![image](https://github.com/user-attachments/assets/4563a643-96be-4238-aef9-67b9fed80fe9)
Here it is with the DCF made by `tkbn.calc_density_compensation_function`:
![image](https://github.com/user-attachments/assets/b4e58f3a-2c52-4445-b5e3-83b60c42bc00)

[mrfSpiral.ipynb.txt](https://github.com/user-attachments/files/17613544/mrfSpiral.ipynb.txt)

The reconstruction image from a real life phantom I posted above was reconstructed via NUFFT of a sliding window, I'm going to see if I can incorporate a sliding window tomorrow.
